### PR TITLE
New event cloning system & timetable cloner (WIP)

### PR DESCRIPTION
**Pull request by ThiefMaster**

### CODE REVIEW

File: indico/MaKaC/conference.py [hunk 1]
```diff
@@ -15,10 +15,10 @@
 # along with Indico; if not, see <http://www.gnu.org/licenses/>.
 
 from itertools import ifilter
-from flask_pluginengine import plugin_context
 from sqlalchemy.orm import lazyload, joinedload, noload
 from werkzeug.urls import url_parse
 
+from indico.modules.events.cloning import LegacyEventCloner as EventCloner  # XXX update code importing it from here
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: XXX?

Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: temporary, will go away once i updated the remaining cloners

http://programmers.stackexchange.com/questions/65467/what-does-xxx-mean-in-a-comment

File: indico/core/db/sqlalchemy/principals.py [hunk 2]
```diff
@@ -432,3 +433,19 @@ def __eq__(self, other):
         else:
             raise ValueError('Unexpected object type {}: {}'.format(type(other), other))
         return db.and_(self.cls.type == other.principal_type, *criteria)
+
+
+def clone_principals(cls, principals):
+    """Clone a list of principals
```
Comment (nop33) [inline indico/core/db/sqlalchemy/principals.py]: I read recently that the conventions for the docstring require a period at the end of the first line: https://www.python.org/dev/peps/pep-0257/#id17

File: indico/core/db/sqlalchemy/principals.py [hunk 3]
```diff
@@ -432,3 +433,19 @@ def __eq__(self, other):
...
+
+    :param cls: the principal type to use (a `PrincipalMixin` subclass)
+    :param principals: a collection of these principals
+    """
```
Comment (nop33) [inline indico/core/db/sqlalchemy/principals.py]: no `:return:`? What's `rv`?

File: indico/modules/events/registration/clone.py [hunk 4]
```diff
@@ -26,65 +28,57 @@
...
-        regs_cloner_enabled = bool(self._find_registration_forms()
-                                   .filter(RegistrationForm.registrations.any(Registration.is_active))
-                                   .count())
-        selected = self.event.getType() == 'meeting'
-        return {
-            'form': (_('Registration forms'), forms_cloner_enabled, False),
-            'registrations': (_('Registrations'), regs_cloner_enabled, selected)
-        }
+    name = 'registration_forms'
+    friendly_name = _('Registration forms')
```
Comment (nop33) [inline indico/modules/events/registration/clone.py]: I know this'll piss you off, but I can't help but notice that you used single quotes for a translatable string! But I know, I know, we don't have a strong convention for that!

File: indico/modules/events/templates/cloner_options.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    </li>
+
+    <script>
+        (function() {
+            'use strict';
+
+            function _update() {
+                var $this = $(this);
+                // note that the data attributes also contain names of internal cloners which will not
+                // have corresponding checkboxes.  take this into account when changing the code below
```
Comment (nop33) [inline indico/modules/events/templates/cloner_options.html]: typo: double whitespace

File: indico/core/db/sqlalchemy/principals.py (hunk 3 above)
Comment (ThiefMaster) [inline indico/core/db/sqlalchemy/principals.py]: `rv` obviously stands for *return value* ;)

File: indico/modules/events/registration/clone.py (hunk 4 above)
Comment (ThiefMaster) [inline indico/modules/events/registration/clone.py]: pff, i never complained about people using single quotes there ;) it's ome doing that!

File: indico/modules/events/templates/cloner_options.html (hunk 5 above)
Comment (ThiefMaster) [inline indico/modules/events/templates/cloner_options.html]: not a typo, double whitespace adds a clearer distinction between sentences

http://english.stackexchange.com/questions/2544/how-many-spaces-should-come-after-a-period-full-stop/2602#2602

> Don't use two spaces **unless you're using a fixed-width font like a typewriter**. If forced to choose only between one space and two, choose one. But if your typesetting system supports it, have a **wider space between sentences**.

Comment (nop33) [inline indico/modules/events/templates/cloner_options.html]: well, that's interesting! Thanks!

File: indico/core/db/sqlalchemy/util/models.py [hunk 6]
```diff
@@ -142,6 +142,18 @@ def populate_from_dict(self, data):
                 raise ValueError("{} has no attribute '{}'".format(cls.__name__, key))
             setattr(self, key, value)
 
+    def populate_from_attrs(self, obj, attrs):
```
Comment (mic4ael) [inline indico/core/db/sqlalchemy/util/models.py]: Isn't it more like `populate_from_obj`?

Comment (ThiefMaster) [inline indico/core/db/sqlalchemy/util/models.py]: Would fit too but to me that name sounds more like something copying all attributes

File: indico/modules/events/contributions/clone.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    def _clone_subcontribs(self, subcontribs):
+        attrs = get_simple_column_attrs(SubContribution)
+        for old_subcontrib in subcontribs:
+            if old_subcontrib.is_deleted:
+                continue
+            subcontrib = SubContribution()
+            subcontrib.populate_from_attrs(old_subcontrib, attrs)
+            subcontrib.references = list(self._clone_references(SubContributionReference, old_subcontrib.references))
+            self._subcontrib_map[old_subcontrib] = subcontrib
+            # TODO: person links
```
Comment (ThiefMaster) [inline indico/modules/events/contributions/clone.py]: those will be cloned once #2257 is merged

File: indico/modules/attachments/clone.py [hunk 8]
```diff
@@ -16,56 +16,78 @@
...
+        with db.session.no_autoflush:
+            self._clone_attachments(new_event)
+        db.session.flush()
 
-    def get_options(self):
-        enabled = bool(self.find_attachments().count())
-        return {'attachments': (_("Materials"), enabled, False)}
+    def _query_folders(self, base_query, for_event):
+        query = (base_query
+                 .filter_by(is_deleted=False)
```
Comment (mic4ael) [inline indico/modules/attachments/clone.py]: Shouldn't you move the call to the `filter_by` after `options`?

Comment (ThiefMaster) [inline indico/modules/attachments/clone.py]: doesn't matter where it is

Comment (mic4ael) [inline indico/modules/attachments/clone.py]: i know, however i remember you commenting on something similar :P (that we should place `filter/filter_by` at the end)

Comment (ThiefMaster) [inline indico/modules/attachments/clone.py]: hm.. can't remember - maybe it was about having all filters calls together?

File: indico/modules/events/templates/cloner_options.html [hunk 9]
```diff
@@ -0,0 +1,42 @@
+{% for cloner in cloners if cloner.is_visible %}
+    {% set available = cloner.is_available -%}
+    <li>
+        <input type="checkbox" name="cloners" id="cloner-{{ cloner.name }}" value="cloner_{{ cloner.name }}"
+               class="js-cloner-option"
+               data-requires="{{ cloner.requires_deep | list | tojson | forceescape }}"
+               data-required-by="{{ cloner.required_by_deep | list | tojson | forceescape }}"
+               {{ 'checked' if available and cloner.is_default }}
+               {{ 'disabled' if not available }}>
+        {#--#}
```
Comment (mic4ael) [inline indico/modules/events/templates/cloner_options.html]: ?

Comment (ThiefMaster) [inline indico/modules/events/templates/cloner_options.html]: swallowing white space

Comment (OmeGak) [inline indico/modules/events/templates/cloner_options.html]: ( ͡° ͜ʖ ͡°)

**Pull request merged**